add json templating

KiaraGrouwstra opened the pull request:
this PR adds JSON templating, following the [outline](https://pad.lassul.us/vars-templates).
for forward compatibility, i exposed the generators' configuration options to the user.

KiaraGrouwstra committed:
add json templating

Lassulus commented on lib.nix:
```diff
@@ -21,4 +21,13 @@
+    text = generator value;
+  };
+
+  template_json = options: template_generator (lib.generators.toJSON options);
+  template_yaml = options: template_generator (lib.generators.toYAML options); # just json
+  template_ini = options: template_generator (lib.generators.toINI options);
```
ah this is an interesting generic solution. I have to check out if it actually supports the way I wanted to use it. My idea involed more jq coding, but this probably fine as well?

KiaraGrouwstra replied:
feedback welcome. ^^
to be fair, i similarly wondered about your implementation going with custom substitution, whereas my earlier attempt had [just settled for `sed`](https://codeberg.org/kiara/dots/src/commit/e72ba33c033438a446efb842a873de805ffd60a3/system/vars.nix#L32). i presumed that to be something about preventing secrets from leaking somehow?

Lassulus replied:
the idea for the custom substition logic, is that the template file is self contained, the python executable is generic, so all information required to do the substitution is inside the template file. I also wanted to change the format a bit, from a simple path to a more complex json structure which can also encode additional information (like quoting/escaping flags). Also i wanted to add another type than just a path. A way to insert a script ececution into the template, but I'm unsure about security implications for now :)

KiaraGrouwstra replied:
fyi i played with this for a bit, didn't get it to work yet but my wip attempt [may be found here](https://github.com/KiaraGrouwstra/nix-templating/compare/json...KiaraGrouwstra:nix-templating:escaping?expand=1)

KiaraGrouwstra replied:
i added the escaping logic as well now. i'm def curious about your thoughts. :smile_cat:

KiaraGrouwstra replied:
added toml as well

Lassulus commented:
code looks pretty nice, I'm currently visiting friends, so please remind me again if I don't merge this until sunday :)

KiaraGrouwstra committed:
escape json

KiaraGrouwstra committed:
add shorthands

KiaraGrouwstra committed:
use more common naming scheme

KiaraGrouwstra committed:
add toml

flesh out toml

move function

restore translations for ini

source escaping rules

tests pass

KiaraGrouwstra committed:
fix arg count

Lassulus commented on lib.nix:
```diff
@@ -1,4 +1,10 @@
 { pkgs, lib, nix_templater }:
+let
+  escapeJson = {
+    "\"" = ''\"'';
+    "\\" = ''\\'';
+  };
+in
```
I'm not sure I understand why we need to escape here?

KiaraGrouwstra replied:
the idea is that if the string you want to insert into json already contained such values, given json already ascribes meaning to these characters it would need to escape them
